### ISSUE TITLE
lax.while_loop within jax.disable_jit() should not silently fall back to the primitive version.

### ISSUE THREAD

**Issue by YouJiacheng** [google/jax#10023]
https://github.com/google/jax/blob/b43ca61a93aec9a638b5fd6b04518869d45201aa/jax/_src/lax/control_flow.py#L279-L288

It should emit a warning or raise an error.

**Comment by jakevdp** [comment]
Can you say more about why you think the behavior is problematic?

**Comment by YouJiacheng** [comment]
Okay. I don't really think this behavior is a bug, thus I use enhancement label.
But I think a good program should not surprise users, i.e. users should be informed if anything doesn't behave as expected, especially when `jax.disable_jit` is intended for debugging.
For example, a user may assume that within `jax.disable_jit`, any "memory leak" should have nothing to do with jit cache, and doesn't realize the "memory leak" is actually not a leak but a cache.

**Comment by YouJiacheng** [comment]
Sure.
Actually it isn't a memory leak, but a cache. (the number of live buffers will stop increasing at 2048)
```python
import jax
from jax import lax
from jax import numpy as jnp

def f(max):
    def cond(val):
        return val < max

    def body(val):
        return val + 1
    return lax.while_loop(cond, body, 0)

def print_num_live_buffers():
    bs = jax.lib.xla_bridge.get_backend().live_buffers()
    print(len(bs))

for _ in range(3):
    x = jnp.full((), 2)
    f(x)
    print_num_live_buffers()
```
Out:
```
1
2
3
```
`disable_jit` works:
```python
import jax
from jax import lax
from jax import numpy as jnp

def f(max):
    def cond(val):
        return val < max

    def body(val):
        return val + 1
    return lax.while_loop(cond, body, 0)

def print_num_live_buffers():
    bs = jax.lib.xla_bridge.get_backend().live_buffers()
    print(len(bs))

for _ in range(3):
    x = jnp.full((), 2)
    with jax.disable_jit():
        f(x)
    print_num_live_buffers()
```
Out:
```
1
1
1
```
`disable_jit` silently doesn't work:
```python
import jax
from jax import lax
from jax import numpy as jnp

@jax.vmap
def f(max):
    def cond(val):
        return val < max

    def body(val):
        return val + 1
    return lax.while_loop(cond, body, 0)

def print_num_live_buffers():
    bs = jax.lib.xla_bridge.get_backend().live_buffers()
    print(len(bs))

for _ in range(3):
    x = jnp.full(1, 2)
    with jax.disable_jit():
        f(x)
    print_num_live_buffers()
```
```
1
2
3
```

**Comment by YouJiacheng** [comment]
@jakevdp friendly ping.
and WDYT about:
https://github.com/google/jax/issues/9714#issuecomment-1079405996
> But there may be an enhancement to have a more flexible cache.

**Comment by shoyer** [comment]
Would it be better to raise a warning or error here?

**Comment by jakevdp** [comment]
I think I'm still confused about what this issue is proposing. Are you saying we should not catch the error here? https://github.com/google/jax/blob/fe035f56fcd68be88999c03d283a079165720d97/jax/_src/lax/control_flow.py#L285-L288

**Comment by YouJiacheng** [comment]
@jakevdp Actually I want to re-raise some warning here, e.g. log the comment to stderr.
